src/thunderbots/software/geom/polygon.cpp

```diff
@@ -0,0 +1,89 @@
+#include "geom/polygon.h"
+
+#include <unordered_set>
+
+#include "geom/util.h"
+
+// CMake defines NDEBUG for Release build configurations
+#ifndef NDEBUG
+#define VALIDATE_POLYGON
+#endif
+
+Polygon::Polygon(const std::vector<Segment>& _segments) : segments(_segments)
+{
+#ifdef VALIDATE_POLYGON
+    if (!isValid())
+    {
+        throw std::invalid_argument("input segments do not form a polygon!");
+    }
+#endif
+}
+
+Polygon::Polygon(const std::initializer_list<Segment>& _segments) : segments(_segments)
+{
+#ifdef VALIDATE_POLYGON
+    if (!isValid())
+    {
+        throw std::invalid_argument("input segments do not form a polygon!");
+    }
+#endif
+}
+
+bool Polygon::containsPoint(const Point& point)
+{
+    // cast a ray from the point in the +x direction
+    Ray ray(point, Vector(1, 0));
+    unsigned int num_intersections = 0;
+    for (const Segment& seg : segments)
+    {
+        // one-definition rule, use anonymous namespace for
+        // intersects() in geom/util.h
+        if (::intersects(ray, seg))
+        {
+            num_intersections++;
+        }
+    }
+    // if the ray intersects the polygon an odd number of times,
+    // it is inside the polygon, otherwise it is outside the polygon
```

Comment:
Could we toss this link here?
https://stackoverflow.com/questions/217578/how-can-i-determine-whether-a-2d-point-is-within-a-polygon

Reply:
added